src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializer.cs

```diff
@@ -12,13 +12,19 @@ namespace System.Runtime.Serialization
     using System.Text;
     using System.Security;
     using DataContractDictionary = System.Collections.Generic.Dictionary<System.Xml.XmlQualifiedName, DataContract>;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Runtime.Serialization.Json;
 
     public abstract class XmlObjectSerializer
     {
+        [RequiresUnreferencedCode(DataContractJsonSerializer.SerializerTrimmerWarning)]
         public abstract void WriteStartObject(XmlDictionaryWriter writer, object? graph);
```

Comment:
This isn't necessarily true. This class is public abstract and may not require unreferenced code. WCF carries an implementation of this abstract type and customers might in their own code too. The WCF implementation does delegate work to XmlSerializer (which doesn't implement this abstract class) and doesn't reference anything from DCS or DCJS.

Reply:
The way the API is shaped - `object? graph` - doesn't lead to being trim compatible. The trimmer has no way of knowing statically which Types are being passed into `WriteStartObject`. Therefore, it doesn't know what members on what Types should be preserved.

Even if you looked at the type of variables:

```C#
Customer c = GetSomeCustomer();
serializer.WriteStartObject(writer, c);
```

You don't know statically if `c` is concretely a `Customer` or a type derived from `Customer`.